src/engine/renderer/glsl_source/cull_cp.glsl

```diff
@@ -73,21 +75,97 @@ struct Plane {
     float distance;
 };
 
+uniform uint u_ViewID;
 uniform uint u_TotalDrawSurfs;
 uniform uint u_SurfaceCommandsOffset;
 uniform vec4 u_Frustum[6]; // xyz - normal, w - distance
+uniform bool u_UseFrustumCulling;
+uniform bool u_UseOcclusionCulling;
+uniform vec3 u_CameraPosition;
+uniform mat4 u_ModelViewMatrix;
+uniform uint u_ViewWidth;
+uniform uint u_ViewHeight;
+uniform float u_P00;
+uniform float u_P11;
+
+// Based on https://zeux.io/2023/01/12/approximate-projected-bounds/
+bool ProjectSphere( in vec3 center, in float radius, in float zNear, in float P00, in float P11, inout vec4 boundingBox ) {
+    if ( -center.z - radius < zNear ) {
+		return false;
+    }
+
+	vec3 cr = center * radius;
+	float czr2 = center.z * center.z - radius * radius;
+
+	float vx = sqrt( center.x * center.x + czr2 );
+	float minx = center.x * center.x + czr2 >= 0.0 ? ( vx * center.x - cr.z ) / ( vx * center.z + cr.x ) : -1.0;
+	float maxx = center.x * center.x + czr2 >= 0.0 ? ( vx * center.x + cr.z ) / ( vx * center.z - cr.x ) : 1.0;
+
+	float vy = sqrt( center.y * center.y + czr2 );
+	float miny = center.y * center.y + czr2 >= 0.0 ? ( vy * center.y - cr.z ) / ( vy * center.z + cr.y ) : -1.0;
+	float maxy = center.y * center.y + czr2 >= 0.0 ? ( vy * center.y + cr.z ) / ( vy * center.z - cr.y ) : 1.0;
+
+	boundingBox = vec4( minx * P00, miny * P11, maxx * P00, maxy * P11 );
+	boundingBox = boundingBox.xwzy * vec4( 0.5f, -0.5f, 0.5f, -0.5f ) + vec4( 0.5, 0.5, 0.5, 0.5 ); // clip space -> uv space
+
+	return true;
+}
 
 bool CullSurface( in BoundingSphere boundingSphere ) {
+    bool culled = false;
+
+    // Frustum culling
     // Skip far plane because we always make it encompass the whole map in the current direction
     // This might need to be changed later for shadowmaps since lights could have some far plane set
     for( int i = 0; i < 5; i++ ) {
         const float distance = dot( u_Frustum[i].xyz, boundingSphere.center ) - u_Frustum[i].w;
 
         if( distance < -boundingSphere.radius ) {
-            return true;
+            culled = u_UseFrustumCulling;
```

Comment:
Since u_UseFrustumCulling is a uniform value couldn't we put the for loop inside an `if`?

Reply:
Right, done now.